File: glsa_parser.go
package main

import (
	"encoding/xml"
	"fmt"
	"log"
	"net/http"
	"strings"
	"sync"
	"os"
	"github.com/PuerkitoBio/goquery"
	"github.com/quay/goval-parser/oval"
	gocvss20 "github.com/pandatix/go-cvss/20"
	gocvss30 "github.com/pandatix/go-cvss/30"
	gocvss31 "github.com/pandatix/go-cvss/31"
	gocvss40 "github.com/pandatix/go-cvss/40"
)

type GLSAParser struct {
	baseURL string
	client  *http.Client
}

func NewGLSAParser(baseURL string) *GLSAParser {
	return &GLSAParser{
		baseURL: baseURL,
		client:  &http.Client{},
	}
}

func (p *GLSAParser) ScrapeGLSAs() error {
	glsaURL := p.baseURL + "/glsa"
	doc, err := p.fetchAndParseHTML(glsaURL)
	if err != nil {
		return err
	}

	glsaLinks, err := extractGLSALinks(doc)
	if err != nil {
		return err
	}

	var wg sync.WaitGroup
	resultChan := make(chan *ovalparser.OvalDefinitions, len(glsaLinks))

	for _, link := range glsaLinks {
		glsaPageURL := p.baseURL + link
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			ovalDefinitions, err := p.parseGLSAPage(url)
			if err != nil {
				log.Printf("Error parsing GLSA page %s: %v", url, err)
				return
			}
			resultChan <- ovalDefinitions
		}(glsaPageURL)
	}

	wg.Wait()
	close(resultChan)

	for ovalDefinitions := range resultChan {
		ovalXML, err := xml.MarshalIndent(ovalDefinitions, "", "  ")
		if err != nil {
			log.Printf("Error generating OVAL XML: %v", err)
			continue
		}
		saveOvalXML(ovalXML)
	}

	return nil
}

func (p *GLSAParser) fetchAndParseHTML(url string) (*goquery.Document, error) {
	resp, err := p.client.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}

	return doc, nil
}

func extractGLSALinks(doc *goquery.Document) ([]string, error) {
	var links []string
	doc.Find("a[href*='/glsa/']").Each(func(i int, s *goquery.Selection) {
		href, exists := s.Attr("href")
		if exists {
			links = append(links, href)
		}
	})
	return links, nil
}

func (p *GLSAParser) parseGLSAPage(glsaPageURL string) (*ovalparser.OvalDefinitions, error) {
	doc, err := p.fetchAndParseHTML(glsaPageURL)
	if err != nil {
		return nil, err
	}

	affectedPackages, err := extractAffectedPackages(doc)
	if err != nil {
		return nil, err
	}

	resolutionSteps, err := extractResolutionSteps(doc)
	if err != nil {
		return nil, err
	}

	cveReferences, err := extractCVEReferences(doc)
	if err != nil {
		return nil, err
	}

	ovalDescription := "Description of the vulnerability"
	ovalDefinitions, err := generateOvalXML(glsaPageURL, affectedPackages, resolutionSteps, cveReferences, ovalDescription)
	if err != nil {
		return nil, err
	}

	return ovalDefinitions, nil
}

// parseCVSS parses a CVSS vector string and returns the parsed CVSS object and its version.
func parseCVSS(vector string) (interface{}, string, error) {
	switch {
	default: // Should be CVSS v2.0 or is invalid
		cvss, err := gocvss20.ParseVector(vector)
		if err != nil {
			return nil, "", err
		}
		return cvss, "2.0", nil
	case strings.HasPrefix(vector, "CVSS:3.0"):
		cvss, err := gocvss30.ParseVector(vector)
		if err != nil {
			return nil, "", err
		}
		return cvss, "3.0", nil
	case strings.HasPrefix(vector, "CVSS:3.1"):
		cvss, err := gocvss31.ParseVector(vector)
		if err != nil {
			return nil, "", err
		}
		return cvss, "3.1", nil
	case strings.HasPrefix(vector, "CVSS:4.0"):
		cvss, err := gocvss40.ParseVector(vector)
		if err != nil {
			return nil, "", err
		}
		return cvss, "4.0", nil
	}
}

// extractCVEReferences extracts CVE references from a given advisory page.
func extractCVEReferences(doc *goquery.Document) ([]oval.CveReference, error) {
	var cveReferences []oval.CveReference

	doc.Find("h3:contains('References')").NextUntil("h3").Find("a[href*='/CVE-']").Each(func(i int, s *goquery.Selection) {
		href, exists := s.Attr("href")
		if !exists {
			return
		}

		cveID := strings.TrimPrefix(strings.TrimPrefix(href, "https://cve.mitre.org/cgi-bin/cvename.cgi?name="), "CVE-")
		cveDescription := s.Text()

		cvss, version, err := parseCVSS(cveDescription)
		if err != nil {
			log.Printf("Error parsing CVSS vector for %s: %v", cveID, err)
			return
		}

		cveReferences = append(cveReferences, oval.CveReference{
			CveID:       cveID,
			Description: cveDescription,
			CvssVersion: version,
			CvssScore:   cvss,
		})
	})

	return cveReferences, nil
}

// Other functions for extracting affected packages, resolution steps,
// generating OVAL XML, and saving OVAL XML go here...

func main() {
	parser := NewGLSAParser("https://security.gentoo.org")
	if err := parser.ScrapeGLSAs(); err != nil {
		log.Fatalf("Error scraping GLSAs: %v", err)
	}
}
[

)
// glsa-oval.xml fix to add glsa-oval-date-24hourtime.xml"
func saveOvalXML(ovalXML []byte) error {
    filename := "glsa-oval.xml"
    return os.WriteFile(filename, ovalXML, 0644)
}
]

File: glsa_parser.py
#!/usr/bin/python3
#
import requests 
import urllib3
# import  bs4 
from bs4 import BeautifulSoup
import re
from datetime import datetime
from vulnlist import VulnList, nvd

def generate_oval_xml(glsa_url, affected_packages, resolution_steps, cve_references, cvss_scores, oval_description):
    """
    Generate OVAL XML from provided security advisory details.
    """
    remediation_resolution = (
        f"Remediation/Resolution\n"
        f"{resolution_steps}"
    )
    
    # Append the remediation resolution to the oval description
    oval_description += f"\n{remediation_resolution}"
    
    cve_details = ""
    for cve_id in cve_references:
        vulnerability = nvd.get_vulnerability(cve_id)
        if vulnerability:
            cve_details += (
                f"<cve_details>\n"
                f"  <cve_id>{vulnerability.cve_id}</cve_id>\n"
                f"  <description>{vulnerability.description}</description>\n"
                f"  <cvss_score>{vulnerability.cvss_score}</cvss_score>\n"
                f"  <severity>{vulnerability.severity}</severity>\n"
                f"</cve_details>\n"
            )

    oval_xml = f"""<?xml version="1.0" encoding="UTF-8"?>
<oval_definitions xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5">
  <metadata>
    <title>Gentoo GLSA {glsa_url}</title>
    <description>{oval_description}</description>
    <timestamp>{datetime.now().isoformat()}</timestamp>
  </metadata>
  <definition id="GLSA-{glsa_url.split('/')[-1]}" version="1">
    <metadata>
      <vendor>GENTOO_SECURITY_ADVISORY</vendor>
      <source>GLSA-{glsa_url.split('/')[-1]}</source>
    </metadata>
    <criteria>
      <criterion>
        <target>affected_packages</target>
        <condition>
          <test_ref>oval:com.gentoo:def:1</test_ref>
        </condition>
      </criterion>
    </criteria>
    <cve_references>
      {cve_details}
    </cve_references>
  </definition>
</oval_definitions>
"""
    return oval_xml

def save_oval_xml(oval_xml, filename='oval.xml'):
    """
    Save OVAL XML to a file.
    """
    with open(filename, 'w') as file:
        file.write(oval_xml)

def parse_glsa_page(glsa_page_url):
    """
    Parse a GLSA page for affected packages, resolution steps, and CVE references.
    """
    try:
        response = requests.get(glsa_page_url)
        response.raise_for_status()
    except requests.RequestException as e:
        print(f"Error fetching GLSA page: {e}")
        return None, None, None
    
    soup = BeautifulSoup(response.text, 'html.parser')

    # Parse affected packages
    vulnlist = VulnList()
    try:
        affected_packages = vulnlist.get_affected_packages(glsa_page_url)
    except Exception as e:
        print(f"Error retrieving affected packages for {glsa_page_url}: {e}")
        affected_packages = []

    # Parse resolution steps
    resolution_tag = soup.find('h3', text='Resolution')
    resolution_steps = ""
    if resolution_tag:
        resolution_steps = resolution_tag.find_next('pre').text.strip()

    # Parse CVE references
    cve_references = []
    references_tag = soup.find('h3', text='References')
    if references_tag:
        for li in references_tag.find_next('ul').find_all('li'):
            cve_url = li.find('a')['href']
            cve_id = cve_url.split('/')[-1]
            cve_references.append(cve_id)

    return affected_packages, resolution_steps, cve_references

def scrape_glsa():
    """
    Scrape GLSA page for security advisories and process them.
    """
    base_url = 'https://security.gentoo.org'
    glsa_url = base_url + '/glsa'

    try:
        response = requests.get(glsa_url)
        response.raise_for_status()
    except requests.RequestException as e:
        print(f"Error fetching GLSA page: {e}")
        return

    soup = BeautifulSoup(response.text, 'html.parser')
    glsa_links = soup.find_all('a', href=re.compile('/glsa/'))

    for link in glsa_links:
        glsa_page_url = base_url + link['href']
        affected_packages, resolution_steps, cve_references = parse_glsa_page(glsa_page_url)

        if affected_packages is None or resolution_steps is None or cve_references is None:
            continue

        print("Affected Packages:")
        for package in affected_packages:
            print(package)
        print("---")

        # Placeholder values for demonstration purposes
        cvss_scores = []
        oval_description = "Description of the vulnerability"

        oval_xml = generate_oval_xml(
            glsa_page_url,
            affected_packages,
            resolution_steps,
            cve_references,
            cvss_scores,
            oval_description
        )

        save_oval_xml(oval_xml)

if __name__ == '__main__':
    scrape_glsa()
